src/smt/optimization_solver.h

```diff
@@ -80,38 +90,350 @@ class Objective
+  }
+};
+
+/**
+ * The goal is of type Integer 
+*/
+template <>
+struct OMTOptimizerImpl<Rational> : OMTOptimizer
+{
+public: 
+  std::pair<OptResult, CVC4::Node> optimize(SmtEngine *parentSMTSolver, CVC4::Node target, ObjectiveType objType) {
+    // linear search for integer goal 
+    // the smt engine to which we send intermediate queries
+    // for the linear search.
+    std::unique_ptr<CVC4::SmtEngine> optChecker; 
+    CVC4::theory::initializeSubsolver(optChecker);
+    CVC4::NodeManager* nm = optChecker->getNodeManager();
+    // we need to be in incremental mode for multiple objectives since we need to
+    // push pop we need to produce models to inrement on our objective
+    optChecker->setOption("incremental", "true");
+    optChecker->setOption("produce-models", "true");
+    // Move assertions from the parent solver to the subsolver
+    std::vector<Node> p_assertions = parentSMTSolver->getExpandedAssertions();
+    for (const Node &e : p_assertions) {
+      optChecker->assertFormula(e);
+    }
+    CVC4::Result intermediateSatResult = optChecker->checkSat();
+    // Model-value of objective (used in optimization loop)
+    CVC4::Node value;
+    if (intermediateSatResult.isUnknown()) {
+      return std::make_pair(OptResult::OPT_UNKNOWN, value);
+    }
+    if (!intermediateSatResult.isSat()) {
+      return std::make_pair(OptResult::OPT_UNSAT, value);
+    }
+    // asserts objective > old_value (used in optimization loop)
+    CVC4::Node increment;
+    CVC4::Kind incrementalOperator = kind::NULL_EXPR; 
+    if (objType == ObjectiveType::OBJECTIVE_MINIMIZE) {
+      // if objective is MIN, then assert optimization_target < current_model_value 
+      incrementalOperator = kind::LT;
+    } else if (objType == ObjectiveType::OBJECTIVE_MAXIMIZE) {
+      // if objective is MAX, then assert optimization_target > current_model_value 
+      incrementalOperator = kind::GT;
+    }
+    // Workhorse of linear search:
+    // This loop will keep incrmenting/decrementing the objective until unsat
+    // When unsat is hit, 
+    // the optimized value is the model value just before the unsat call
+    while (intermediateSatResult.isSat()) {
+      value = optChecker->getValue(target);
+      Assert(!value.isNull());
+      increment = nm->mkNode(incrementalOperator, target, value);
+      optChecker->assertFormula(increment);
+      intermediateSatResult = optChecker->checkSat();
+    }
+    return std::make_pair(OptResult::OPT_OPTIMAL, value);
+  }
+  virtual ~OMTOptimizerImpl() {}
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) override {
+    return this->optimize(parentSMTSolver, target, ObjectiveType::OBJECTIVE_MINIMIZE);
+  }
+  virtual std::pair<OptResult, CVC4::Node> maximize(SmtEngine *parentSMTSolver, CVC4::Node target) override {
+    return this->optimize(parentSMTSolver, target, ObjectiveType::OBJECTIVE_MAXIMIZE);
+  }
+};
+
+
+template <>
```

Comment:
I am wondering why do we need templates here? Won't we have only 3 types or something like that? Maybe we can keep it simple?